### PR TITLE
JavaScript evaluation instead of Janino for conditional processing

### PR DESCRIPTION
**Pull request by smurf667**

### CODE REVIEW

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/IfAction.java [hunk 1]
```diff
@@ -47,7 +47,10 @@ public void begin(InterpretationContext ic, String name, Attributes attributes)
         }
 
         ic.pushObject(this);
-        if (!EnvUtil.isJaninoAvailable()) {
+        
+        boolean useJavaScript = Boolean.TRUE.equals(ic.getObjectMap().get(PropertyEvalJavaScriptBuilder.CONDITIONAL_JS_ATTR));
```
Comment (shaehnel) [inline logback-core/src/main/java/ch/qos/logback/core/joran/conditional/IfAction.java:51]: can ic.getObjectMap() be null?

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public static final String CONDITIONAL_JS_ATTR = "conditional-js";
+    private static final String JS_ENGINE_NAME = "JavaScript";
+
+    // variable names provided in JavaScript context
+    static final String PROPERTIES = "props";
+    static final String CONTEXT = "ctx";
+
+    private static WeakReference<ScriptEngine> globalEngine = new WeakReference<ScriptEngine>(null);
+    private Context context;
+    ScriptEngine engine;
```
Comment (shaehnel) [inline logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java:28]: default access modifier for easier testing?

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        this.localPropContainer = localPropContainer;
+        
+        engine = globalEngine.get();
+        if (engine == null) {
+            final ScriptEngineManager factory = new ScriptEngineManager();
+            engine = factory.getEngineByName(JS_ENGINE_NAME);
+            if (engine == null) {
+                throw new IllegalStateException("JavaScript not supported");
+            }
+            final Reader reader = new InputStreamReader(getClass().getResourceAsStream("functions.js"));
```
Comment (shaehnel) [inline logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java:42]: curious: why is functions.js loaded?

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    private static class JavaScriptCondition implements Condition {
+        
+        final ScriptEngine executor;
+        final String script;
+
+        JavaScriptCondition(ScriptEngine engine, PropertyContainer properties, Context context, String expression) {
+            executor = engine;
+            Bindings bindings = executor.getBindings(ScriptContext.ENGINE_SCOPE);
+            bindings.put(PROPERTIES, properties);
+            bindings.put(CONTEXT, context);
```
Comment (shaehnel) [inline logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java:77]: it seems that context can be null if setContext is not called. is that okay?

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/IfAction.java (hunk 1 above)
Comment (smurf667) [reply logback-core/src/main/java/ch/qos/logback/core/joran/conditional/IfAction.java:51]: No, the map is created in the constructor of the `InterpretationContext`

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java (hunk 2 above)
Comment (smurf667) [reply logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java:28]: Yes, I wanted to check that different "if" conditions re-use the same engine. I found logback to be somewhat liberal with the scope of some member variables, so I followed along.

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java (hunk 3 above)
Comment (smurf667) [reply logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java:42]: It supplies the actual property accessor functions that are also available via Janino (like `p()` and `isNull()` etc.). Unfortunately, I forget to check that file in. Shame on me! Doing it tonight.

File: logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java (hunk 4 above)
Comment (smurf667) [reply logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java:77]: It is always set, see https://github.com/smurf667/logback/blob/JavaScriptEvaluation/logback-core/src/main/java/ch/qos/logback/core/joran/conditional/IfAction.java#L71

File: logback-core/src/main/resources/ch/qos/logback/core/joran/conditional/functions.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+var OptionHelper = Java.type("ch.qos.logback.core.util.OptionHelper");
+
+var property = function(key) {
+    var val = OptionHelper.propertyLookup(key, props, ctx);
```
Comment (shaehnel) [inline logback-core/src/main/resources/ch/qos/logback/core/joran/conditional/functions.js:4]: where is ctx defined? is it implicitly injected?

Comment (smurf667) [reply logback-core/src/main/resources/ch/qos/logback/core/joran/conditional/functions.js:4]: See https://github.com/smurf667/logback/blob/JavaScriptEvaluation/logback-core/src/main/java/ch/qos/logback/core/joran/conditional/PropertyEvalJavaScriptBuilder.java#L75-L77

File: logback-core/src/main/resources/ch/qos/logback/core/joran/conditional/functions.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    return val ? val : "";
+};
+var p = property;
+
+var isDefined = function(key) {
+    var val = OptionHelper.propertyLookup(key, props, ctx);
+    return val ? true : false;
```
Comment (shaehnel) [inline logback-core/src/main/resources/ch/qos/logback/core/joran/conditional/functions.js]: can this be written as `return !(p(key)=="")`?

**Pull request closed without merging**